x/programs/examples/storage/storage.go

```diff
@@ -0,0 +1,222 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package storage
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	smath "github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/crypto/ed25519"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+const (
+	HRP           = "simulator_key_"
+	programPrefix = 0x0
+	balancePrefix = 0x1
+
+	BalanceChunks uint16 = 1
+)
+
+var (
+	ErrInvalidBalance = errors.New("invalid balance")
+	balanceKeyPool    = sync.Pool{
+		New: func() any {
+			return make([]byte, 1+ed25519.PublicKeyLen+consts.IDLen+consts.Uint16Len)
+		},
+	}
+)
+
+//
+// Program
+//
+
+func ProgramKey(id ids.ID) (k []byte) {
+	k = make([]byte, 1+consts.IDLen)
+	// convert uint64 to bytes
+	k[0] = programPrefix
+	copy(k[1:], id[:])
+	return
+}
+
+// [programID] -> [programBytes]
+func GetProgram(
+	db state.Immutable,
+	programID ids.ID,
+) (
+	[]byte, // program bytes
+	bool, // exists
+	error,
+) {
+	k := ProgramKey(programID)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	v, err := db.GetValue(ctx, k)
+	if errors.Is(err, database.ErrNotFound) {
+		return nil, false, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	return v, true, nil
+}
+
+// SetProgram stores [program] at [programID]
+func SetProgram(
+	db database.KeyValueWriter,
+	programID ids.ID,
+	program []byte,
+) error {
+	k := ProgramKey(programID)
+	v := make([]byte, len(program))
+	copy(v, program)
+	return db.Put(k, v)
+}
+
+//
+// Balance
+//
+
+// [accountPrefix] + [address] + [asset]
+func BalanceKey(pk ed25519.PublicKey, asset ids.ID) (k []byte) {
+	k = balanceKeyPool.Get().([]byte)
+	k[0] = balancePrefix
+	copy(k[1:], pk[:])
+	copy(k[1+ed25519.PublicKeyLen:], asset[:])
+	binary.BigEndian.PutUint16(k[1+ed25519.PublicKeyLen+consts.IDLen:], BalanceChunks)
+	return
+}
+
+// If locked is 0, then account does not exist
+func GetBalance(
+	ctx context.Context,
+	im state.Immutable,
+	pk ed25519.PublicKey,
```

Comment:
We will eventually replace this with a much more generic address scheme but this is ok for now.